### ISSUE TITLE
Tagged unions support

### ISSUE THREAD

**Issue by mariusmuja** [microsoft/pyright#2160]
**Is your feature request related to a problem? Please describe.**

The following code uses tagged unions with `Literal` integer as discriminant:

```python
from typing import Tuple, Union, Literal

MsgA = Tuple[Literal[1], str]
MsgB = Tuple[Literal[2], float]

Msg = Union[MsgA, MsgB]

def handle_a(a: MsgA):
    pass

def handle_b(b: MsgB):
    pass


def handle(m: Msg):
    if m[0] == 1:
        handle_a(m)
    elif m[0] == 2:
        handle_b(m)
```

This type checks successfully with `mypy` as documented here (https://mypy.readthedocs.io/en/stable/literal_types.html#tagged-unions), however this type narrowing pattern is not supported by pyright.

**Describe the solution you'd like**

It would be great if pyright would also support this type narrowing pattern.

**Comment by erictraut** [comment]
This feature will be included in the next release of pyright and pylance.

**Comment by mariusmuja** [comment]
Thank you for adding this feature so quickly! 

With this I can now remove lots of `if TYPE_CHECKING:` hacks from the code.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.161, which I just published. It will also be included in the next release of pylance.
